### PR TITLE
Pretty print

### PR DESCRIPTION
**Pull request by EmilySillars**
Addressing issue #51 
To add indentation/make more specialized formatting for different AST nodes, I added another layer of indirection between the IR and the pretty printer output:
```
src/IR.hs:         indentPretty   

                        |
                        |
                        v

src/IR/IR.hs:           ws       // Whitespace type class

                        |
                        |
                        v

src/IR/IR.hs:         pretty     // Pretty type class
```
The middle layer sort of acts as a more adjustable wrapper for the pretty function. I'm not convinced a typeclass is definitely the right tool here, but I am using that for now and can adjust the code structure as I get a better idea of what is needed.
Currently:
- hello1.ssl can be exported and fed back into the compiler successfully, and looks much more readable.
-  Input hello1.ssl program:
```
main ( cout : & Int ) -> () =
  after 10 , (cout : & Int) <- 72
  wait (cout : & Int)
  after 10 , (cout : & Int) <- 101
  wait (cout : & Int)
  after 10 , (cout : & Int) <- 108
  wait (cout : & Int)
  after 10 , (cout : & Int) <- 108
  wait (cout : & Int)
  after 10 , (cout : & Int) <- 111
  wait (cout : & Int)
  after 10 , (cout : & Int) <- 10
  wait (cout : & Int)
```
- output pretty print using `stack exec -- sslc --dump-ir-final tests/hello1.ssl`
```
main cout : (&Int) -> () = 
 after 10, cout <- 72
 wait cout
 after 10, cout <- 101
 wait cout
 after 10, cout <- 108
 wait cout
 after 10, cout <- 108
 wait cout
 after 10, cout <- 111
 wait cout
 after 10, cout <- 10
 wait cout
```

Next steps:
- prettify match statements with indentation
- make sure nested match, lambdas, and lets are indented properly
- add support for pretty printing type definitions
- solve issue where pretty printer spits out constructor functions starting with "__" which cannot be parsed
- double check annotation on/off switch works as expected and can be fed back in

### CODE REVIEW

**Comment by EmilySillars** [comment]
Update:
- match and nested match statements are pretty
- ADT type definitions are pretty
- __functions can be removed from pretty printing via flag
- map.ssl and adt-recursive.ssl are pretty and can be successfully fed back in to compiler ( __funcs OFF)

Next Steps:
- fix issue with nested lambdas and __funcs ON
- make sure type annotated output can be fed back in
- make sure rest of examples look pretty and can be fed back in
- double check nested lambdas look pretty and can be fed back in
- refactor code/use better function naming

Example input/output of pretty printer:
Original:
```
type List
  Cons Int List
  Nil

print_list2 (out: &Int) (l: List) -> () =
  match l
    Cons i l0 =
      after 1, out <- i
      wait out
      after 1, out <- 32
      wait out
      print_list2 out l0
    Nil = ()

print_list (out: &Int) (l: List) -> () =
    after 1, out <- 91
    wait out
    after 1, out <- 32
    wait out

    print_list2 out l

    after 1, out <- 93
    wait out
    after 1, out <- 10
    wait out

map (f: Int -> Int) (l: List) -> List =
  match l
    Cons i l0 = Cons (f i) (map f l0)
    Nil = Nil

add (a: Int) (b: Int) -> Int =
  a + b

main (out: &Int) -> () =
  let l1 = (Cons 104 (Cons 101 (Cons 108 (Cons 108 (Cons 111 Nil)))))
  print_list out l1
  let l2 = map (add 1) l1
  let l3 = map (add 4) l1
  print_list out l1
  print_list out l2
  print_list out l3
```
Output without __funcs:
```
type List 
 Cons Int (List)
 Nil 

print_list2 (out : &Int) (l : (List)) -> () = 
 match l
       Cons i l0 = after 1, out <- i
                   wait out
                   after 1, out <- 32
                   wait out
                   print_list2 out l0
       Nil  = ()

print_list (out : &Int) (l : (List)) -> () = 
 after 1, out <- 91
 wait out
 after 1, out <- 32
 wait out
 print_list2 out l
 after 1, out <- 93
 wait out
 after 1, out <- 10
 wait out

map (f : Int -> Int) (l : (List)) -> (List) = 
 match l
       Cons i l0 = Cons (f i) (map f l0)
       Nil  = Nil

add (a : Int) (b : Int) -> Int = 
 a + b

main (out : &Int) -> () = 
 let l1 = Cons 104 (Cons 101 (Cons 108 (Cons 108 (Cons 111 (Nil)))))
 print_list out l1
 let l2 = map (add 1) l1
 let l3 = map (add 4) l1
 print_list out l1
 print_list out l2
 print_list out l3
```
Output with __funcs:
```
type List 
 Cons Int (List)
 Nil 

print_list2 = fun out fun l match l
      Cons i l0 = after 1, out <- i
                  wait out
                  after 1, out <- 32
                  wait out
                  print_list2 out l0
      Nil  = ()

print_list = fun out fun l after 1, out <- 91
wait out
after 1, out <- 32
wait out
print_list2 out l
after 1, out <- 93
wait out
after 1, out <- 10
wait out

map = fun f fun l match l
      Cons i l0 = __Cons (f i) (map f l0)
      Nil  = Nil

add = fun a fun b a + b

main = fun out let l1 = __Cons 104 (__Cons 101 (__Cons 108 (__Cons 108 (__Cons 111 (Nil)))))
print_list out l1
let l2 = map (add 1) l1
let l3 = map (add 4) l1
print_list out l1
print_list out l2
print_list out l3

__Cons = fun __arg0 fun __arg1 Cons __arg0 __arg1
```

File: src/IR/IR.hs
Comment (j-hui) [inline src/IR/IR.hs:357]: Could you change this instance to `instance Pretty (Expr ())`? I want the fact that type annotations are being omitted to be visible in the type signature

File: src/IR/IR.hs
Comment (j-hui) [inline src/IR/IR.hs:433]: `FMT` should be stylized as `Fmt`

But also, this looks like what `Pretty` should actually be? Is there a reason to introduce another typeclass?

File: src/IR/IR.hs
Comment (j-hui) [inline src/IR/IR.hs:465]: Where does the hard-coded indentation level of `6` come from?

**Review by j-hui — COMMENTED** [review]
I left some questions here; let's talk more about this in our meeting later!

**Comment by EmilySillars** [comment]
**Update: Incorporated PR feedback**
- removed FMT typeclass and moved all pretty printing changes to the pretty typeclass
- preserved old pretty printer with type checking in the Lengthy typeclass
- updated spaghetti to call lengthy instead of pretty
- tightened up code in DConToFunc.hs/added more documenation
- replaced magic numbers with indentNo constant defined in Pretty.hs
- updated/improved comments

**Next Steps:**
- I think this PR is ready to merge into main now
- Implementing minimum parentheses algo or selective type annotations could be good work for future PRs
- I think the original pretty printer may have only ever passed unit tests? Maybe we can talk more about this at the meeting.

**Review by j-hui — CHANGES_REQUESTED** [review]
I'd like to disentangle `Pretty` from `Lengthy` (or `Dumpy`) entirely if at all possible. In particular, we will end up with three typeclasses for converting ASTs (or IRs) to Strings (or Docs, and then eventually Strings):

- `Pretty`: Show the prettiest-possible, formatted representation of the AST, with indentation, minimal parens, etc.. This doesn't have to be exactly what the user wrote, but it should more or less look like what a human would write. `Pretty` may end up _losing_ a lot of information (in particular, type information), but that's ok. The point is to produce something readable.
- `Lengthy`/`Dumpy`: Show an ugly but parseable representation of the AST. In particular, it should be able to be parsed by our parser without the lexer's layout algorithm having to insert anything. I'm not sure how useful this will end up being, but it should at least give us a _stable_, machine-parseable representation of the program (i.e., doesn't depend on what way we define the AST/IR data types).
- `Show`: Used to dump out the Haskell ADT definition as `String`. This is used to show every last bit of information that's present in some AST/IR, at the cost of readability. This is mainly going to be useful when the `Pretty` instance fails us. We should always derive this automatically rather than hand-write this, so that it is compatible with automatically-derived `Read` instances (i.e., `read . show === id`).

We shouldn't have any typeclass-level dependencies between these three typeclasses (i.e., there shouldn't be `Pretty t => Lengthy t` or anything like that). One instance may defer to another to avoid code repetition, but that shouldn't be baked into the definition or instance declaration of any typeclass. For example, the `Pretty Literal` instance may just defer to `Lengthy Literal` like this:

```haskell
instance Pretty Literal where
  pretty  = lengthy
```

But only for trivial data types where we know the two are represented exactly the same, independent of readability concerns.

File: src/IR/IR.hs
Comment (EmilySillars) [reply src/IR/IR.hs:371]: Yes, I tried to do that but I was getting a type error:
```
• Illegal instance declaration for ‘Pretty (Expr ())’
    (All instance types must be of the form (T a1 ... an)
     where a1 ... an are *distinct type variables*,
     and each type variable appears at most once in the instance head.
     Use FlexibleInstances if you want to disable this.)
```
I will enable flexible instances and make this pretty instead of fmtExpr.

File: src/Common/Pretty.hs
Comment (EmilySillars) [reply src/Common/Pretty.hs:71]: When you say that Lengthy should not depend on Pretty, this means I should be defining instances of Dumpy for literals like Int, right? For example, should I be defining something like this:
`instance Dumpy Int where dumpy = pretty`?
Or should it look like
`instance Dumpy Int where dumpy = unsafeViaShow`?

File: src/IR/Types/Classes.hs
Comment (XijiaoLi) [reply src/IR/Types/Classes.hs:42]: For sure, feel free to change it!

File: src/Common/Pretty.hs
Comment (j-hui) [reply src/Common/Pretty.hs:71]: I mean the class declaration should look like:

```haskell
class Lengthy a where
  lengthy :: a -> Doc
```

The specific instance may use `pretty` to implement `lengthy` (or vice versa), but that dependency should not be required in the typeclass signature

**Comment by EmilySillars** [comment]
Hi John,
     - I believe I have completely separated Dumpy and Pretty per your request
     - I'm not sure about the change you requested on ast.hs; could you respond to my question on what to do for this part?
     - while updating dumpy to work with the regression tests, I found a bug in the dConToFunc.hs pass which I fixed
     - without pattern desugaring (Leo's pass), all the regression tests work with dumpy, however when I test dumpy with pattern desugaring, map.ssl,map-inc.ssl and adt-partially-applied.ssl all fail. I was trying to resolve this issue this week, but I was getting stuck and thought I better check in with you.
    I realize the semester is over now so maybe I shouldn't be working on this, but I remember you saying you wanted to merge in the pretty-print branch soon and I didn't want to leave this PR in an unfinished state for you. If you'd prefer I stop working on this and you can take over from here, let me know. Otherwise I will keep working on getting dumpy to pass all regression tests. Let me know your thoughts/ what you would like me to do!

File: src/Common/Pretty.hs
Comment (j-hui) [inline src/Common/Pretty.hs:22]: "constrained by Lengthy" -> "constrained by 'Dumpy'"

File: src/IR.hs
Comment (j-hui) [inline src/IR.hs:50]: Just use `Bool` as the type here, no need for the type alias

File: src/IR/DConToFunc.hs
Comment (j-hui) [inline src/IR/DConToFunc.hs:121]: You should only really use `-- ^` for things like data constructors, where you document the item inline

File: src/IR/DConToFunc.hs
Comment (j-hui) [inline src/IR/DConToFunc.hs:157]: I would argue that the user should not be able to write this sslang program

File: src/IR/DConToFunc.hs
Comment (j-hui) [inline src/IR/DConToFunc.hs:179]: This seems like it would actually break programs that define functions that happen to clash with compiler-generated functions. Those functions should not be allowed in the first place.

File: src/IR/IR.hs
Comment (j-hui) [inline src/IR/IR.hs:2]: Which definition requires this?

**Review by j-hui — APPROVED** [review]
This looks good as it is. Let's merge this for now, and we can continue tweaking things as we run into problems.

Thanks for the great work @EmilySillars !

File: src/IR/DConToFunc.hs
Comment (EmilySillars) [reply src/IR/DConToFunc.hs:179]: I agree that users should not be able to write functions that clash with compiler-generated functions. I remember we changed the scanner to allow users to begin function names with underscores around April or early May. Starting functions with underscores was the only way I knew of to distinguish compiler generated vs user defined functions. 
1) Would you like me to revert that change to the scanner? The issue here is if we do change the scanner, we can no longer read back in the output of the pretty printer, because dcontofunc generates functions that start with underscores. 
2) Alternatively, maybe dcontofunc could output its constructor functions to a different file that gets linked in. Then the output of the pretty printer could be tested (and fed back in to the front of the compiler), and if any functions with name collisions are detected by dcontofunc in the original source, it could throw an error.

File: src/IR/IR.hs
Comment (EmilySillars) [reply src/IR/IR.hs:2]: `instance (Pretty Expr ()) where` (line 370) requires this.
Without it, I get the error:
```
• Illegal instance declaration for ‘Pretty (Expr ())’
    (All instance types must be of the form (T a1 ... an)
     where a1 ... an are *distinct type variables*,
     and each type variable appears at most once in the instance head.
     Use FlexibleInstances if you want to disable this.)
• In the instance declaration for ‘Pretty (Expr ())’typecheck
```
It doesn't like that I'm only implementing the Pretty typeclass for `Expr t `when `Expr t`'s type variable is unit.

File: src/IR/DConToFunc.hs
Comment (EmilySillars) [reply src/IR/DConToFunc.hs:157]: **Would you like me to address this issue before merging this PR into main?** I could add a check in `dcontofunc` that throws an error on finding name collisions with user defined functions, but this will break our pretty printer tests. I talked a bit more about this here: (https://github.com/ssm-lang/sslang/pull/92#discussion_r907892990)

File: src/IR/IR.hs
Comment (j-hui) [reply src/IR/IR.hs:2]: Ah right, got it!

File: src/IR/DConToFunc.hs
Comment (j-hui) [reply src/IR/DConToFunc.hs:178]: Let's fix this later! For now, I want to finalize this PR so that I can make some destructive changes on the IR (:

**Pull request merged**
